### ISSUE TITLE
Issue running pipeline on server

### ISSUE THREAD

**Issue by VSack** [hzi-bifo/art_illumina_profiles#2]
I just tried to run the pipeline on the server and got the following error:
```
CreateCondaEnvironmentException:
Could not create conda environment from /net/sgi/singlecell/art-profiles/art_illumina_profiles/envs/wget.yaml:
Fetching package metadata .............

NoPackagesFoundError: Package missing in current linux-64 channels: 
  - wget 1.19.4*
```
I set up the channels as described in the Readme/the bioconda page . Any idea where this stems from?

**Comment by dlaehnemann** [comment]
That is a funny one, because it should be there. With the bioconda setup, if I do a `conda search wget`, I get:
```
Loading channels: done
# Name                  Version           Build  Channel             
wget                       1.18               0  pkgs/free           
wget                     1.19.1      he4ec0ba_0  pkgs/main           
wget                     1.19.4      hf388c53_0  pkgs/main           
wget                     1.19.5      h1ad7b7a_0  pkgs/main           
wget                     1.19.5      h20c2e04_0  pkgs/main 
```

Could you double-check that? Also, could you double-check your channel setup with `conda config --show-sources`, I get:

```
==> /home/david/.condarc <==
channels:
  - conda-forge
  - bioconda
  - defaults
```

Another possibility are different `conda` versions, because I think they changed channel names a while ago. Could you check `conda --version` to give `conda 4.5.12`? Otherwise, Update it with `conda update -n base conda` and also in the `snakemake` environment with `conda update -n snakemake conda`.

If none of the above helps, just ping me again...

**Comment by VSack** [comment]
It seems like the version was a problem. After I updated to `conda 4.5.12`, I deleted and set up the snakemake environment from scratch. However, now I'm getting a new error:

```
...

Activating conda environment: /net/sgi/singlecell/art-profiles/art_illumina_profiles/.snakemake/conda/ce368734
Activating conda environment: /net/sgi/singlecell/art-profiles/art_illumina_profiles/.snakemake/conda/ce368734
processing /tmp/tmp85sstk3v/input.fq
already processed the file /tmp/tmp3dt2qk8g/input.fq
remove input.fq.txt to re-process /tmp/tmp3dt2qk8g/input.fq
The read profile file profiles/test.1.txt has been created
Traceback (most recent call last):
  File "/net/sgi/singlecell/art-profiles/art_illumina_profiles/.snakemake/scripts/tmpou6cjc6n.wrapper.py", line 47, in <module>
    "( art_profiler_illumina {snakemake.params} {profile_name}"
  File "/home/vsack/miniconda2/envs/snakemake/lib/python3.6/site-packages/snakemake/shell.py", line 142, in __new__
    raise sp.CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command ' set -euo pipefail;  ( art_profiler_illumina  profiles/test.2 /tmp/tmp85sstk3v fq 2 ) 2> logs/art_profiler_illumina/test.2.log ' returned non-zero exit status 1.
[Fri Jan 18 10:13:38 2019]
Finished job 1.
3 of 5 steps (60%) done
[Fri Jan 18 10:13:38 2019]
Error in rule art_profiler_illumina:
    jobid: 2
    output: profiles/test.2.txt
    log: logs/art_profiler_illumina/test.2.log
    conda-env: /net/sgi/singlecell/art-profiles/art_illumina_profiles/.snakemake/conda/ce368734

RuleException:
CalledProcessError in line 43 of /net/sgi/singlecell/art-profiles/art_illumina_profiles/Snakefile:
Command 'source activate /net/sgi/singlecell/art-profiles/art_illumina_profiles/.snakemake/conda/ce368734; set -euo pipefail;  /home/vsack/miniconda2/envs/snakemake/bin/python3.6 /net/sgi/singlecell/art-profiles/art_illumina_profiles/.snakemake/scripts/tmpou6cjc6n.wrapper.py ' returned non-zero exit status 1.
  File "/net/sgi/singlecell/art-profiles/art_illumina_profiles/Snakefile", line 43, in __rule_art_profiler_illumina
  File "/home/vsack/miniconda2/envs/snakemake/lib/python3.6/concurrent/futures/thread.py", line 56, in run
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /net/sgi/singlecell/art-profiles/art_illumina_profiles/.snakemake/log/2019-01-18T101049.831035.snakemake.log
```
(I should note here that I'm still trying to run this on the small example dataset specified in the config, not on the big one yet - that should not influence the error though.)

**Comment by dlaehnemann** [comment]
What do the logs say? I.e. the `Complete log` given in the last line of output above, and the log file of the rule that failed (I.e. `logs/art_profiler_illumina/test.2.log`).

**Comment by dlaehnemann** [comment]
If these don't help, it might have to do with the miniconda version, which has been 3 for a while now. So removing the old one and reinstalling the new one from scratch mightbee something to try. But I rather think, it's something else -- the updating of conda should have done the the same as the upgrade to miniconda3, I think...

**Comment by VSack** [comment]
In the art-log:
```
mv: cannot stat ‘input.fq.txt’: No such file or directory
No such file or directory at /net/sgi/singlecell/art-profiles/art_illumina_profiles/.snakemake/conda/ce368734/bin/combinedAvg.pl line 18.
```

In fact, the files currently in the folderstructure are
```
.../art_illumina_profiles$ ls -R
.:
config.yaml  envs  input  LICENSE  logs  profiles  README.md  reads  Snakefile

./envs:
wget.yaml

./input:
test.bam  test.bam.bai

./logs:
art_profiler_illumina  download_bam_bai

./logs/art_profiler_illumina:
test.1.log  test.2.log

./logs/download_bam_bai:
test.log

./profiles:
test.1.txt

./reads:
test.1.fq  test.2.fq
```
To me, it looks a lot like an issue with wildcards, although looking at the snakemake file I'm unsure of the cause.

**Comment by dlaehnemann** [comment]
I think I found the problem -- art_profiler_illumina creates temporary output files in the output directory, and as the output directory is not unique per execution of the rule, we got a name clash there. Now the temporary file names are also unique (and prevent unwanted art file naming magic, as before). It's already in a pull request to the wrapper repo, and will be available once it is merged and released:
<https://bitbucket.org/snakemake/snakemake-wrappers/pull-requests/60/art_profiler_illumina-fix-file-naming-bug/diff>

For now, let's see if the pipeline runs through without the parallelism.

**Comment by VSack** [comment]
Here's a little update from the big-dataset-testrun:

1. The pipeline exited without any message after downloading the bam and the bam.bai-file. The files were there but not the ones of the following steps and there was no process running anymore. I had to unlock the folder to restart the pipeline to make it execute the next step.
2. The rule `rule samtools_bam2fq_separate` needs to be run with one core only, otherwise it will try to produce multiple temporary files and return errors like 
```
samtools sort: failed to create temporary file "reads/RMNISTHS_30xdownsample.1.0000.bam": File exists
```
I'm running the pipeline on one core now, let's see how long that will take...

**Comment by dlaehnemann** [comment]
Hmm, this sounds like an unexpected exit of the pipeline, not caused by the pipeline but somewhere on the system side. The process could have been killed automatically due to some excessive ressource usage, although that sounds unlikely with the given `samtools sort` defaults. Another option is that your session got killed, somehow? Did you run the pipeline from a [`screen`](http://aperiodic.net/screen/quick_reference) session that you logged out from -- that's what I have gotten into the habit of doing, because it gives me consistent server sessions that I can log on to from different places, as well? Or did you run it in the background with `&`, because then it will terminate once you end your server session, unless you ran it with `nohup`. Here's a quick overview of the behaviour:
https://www.maketecheasier.com/run-bash-commands-background-linux/

And then the existing temporary file was probably still around from the previous aborted run of the pipeline. As it is not part of the input / output tracking of snakemake and as the pipeline exited uncontrolled, this was not removed automatically. Does this make sense, or am I missing something (like you having tried with multiple cores after cleaning up the temporary files)? I'm just trying to figure out if there really is another bug in the workflow, or if there is another error source, here.

In any case, keeping the pipeline running on one core seems like a good intermediate solution. Only if that's too slow, consider stopping it, cleaning up all intermediate data manually and then restarting with multiple cores.

**Comment by dlaehnemann** [comment]
Just tried out the pipeline on a larger bam file I had lying around, and the multithreaded `samtools sort -n` in the bam2fq rule ran through without problems. Maybe this problem is also caused by the user problem we have identified, which could have to do with permissions on different storage systems, as the /tmp directory will probably be somewhere separate...

**Comment by dlaehnemann** [comment]
This issue should be resolved:

1. The problem with parallelism in `art_profiler_illumina` is resolved via [changes to the wrapper that have already been merged](https://bitbucket.org/snakemake/snakemake-wrappers/pull-requests/60/art_profiler_illumina-fix-file-naming-bug/diff). They will be pulled into any run of the pipeline automatically, once there is a new release of the Snakemake wrappers.
2. The problem with parallelism in `samtools sort -n` in the `samtools bam2fq` rule seems to be specific to your user configuration or to some `/net/sgi` configuration -- as mentioned above, I tested the pipeline with a different (large) bam file on different storage and the rule ran through perfectly.

So I'll close this for now. If we encounter further problems in the future, we'll just open new issues.
